### PR TITLE
Adds neil.el - Emacs companion

### PR DESCRIPTION
**Pull request by agzam**
@borkdude I'm not sure if anyone has already started something like this.

~This is an early draft; I'm hoping to add a few more features and perhaps publish it (at least on MELPA). WDYT?~

Performs the search, optionally lets you choose a version; Optional automatic in-project dependency injection.

### CODE REVIEW

**Comment by borkdude** [comment]
I think it's a very cool idea! Please pursue this :).

**Comment by agzam** [comment]
Alright. I just wasn't sure if someone already made one. I'll try cleaning it up and make a proper PR. Thank you!

**Comment by agzam** [comment]
@borkdude the basics are in - performs a library search, lets you choose from versions; optionally adds the dependency to deps.edn.

**Comment by borkdude** [comment]
@agzam Should the copyright stuff be updated as well? Or if you insist, you can maintain it in your own repo? I don't know how it technically works if there is a copyright statement with a different author in one of the repos I make changes to.

File: README.md [hunk 1]
```diff
@@ -181,7 +181,28 @@ NOTE: invoking a clojure tool requires you to quote strings:
 clj -Tneil add-dep :lib org.clojure/tools.cli :version '"1.0.206"'
 ```
 
-## Roapmap
+## Emacs Integration 
+
+[neil.el](https://github.com/babashka/neil/blob/main/neil.el) is a companion Emacs package. 
+
+Load it using your proffered Emacs package manager, e.g., for Doom Emacs:
```
Comment (borkdude) [inline README.md:188]: preferred?

**Comment by borkdude** [comment]
I think adding to deps.edn should be enabled by default and can be optionally disabled. What else is the point for searching for a dep? :)

**Comment by borkdude** [comment]
Works great btw: https://twitter.com/borkdude/status/1517810425689219072

File: neil.el [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+;;
+;; Copyright (C) 2022 Ag Ibragimov
+;;
+;; Author: Ag Ibragimov <agzam.ibragimov@gmail.com>
+;; Maintainer: Ag Ibragimov <agzam.ibragimov@gmail.com>
+;; Created: April 20, 2022
+;; Modified: April 20, 2022
+;; Version: 0.0.1
+;; Keywords: convenience tools
+;; Homepage: https://github.com/agzam/neil
```
Comment (teodorlu) [inline neil.el:11]: Homepage is fork repo?

File: neil.el [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+              (seq-map
+               (lambda (s)
+                 (when (string-match ":lib \\(.*\\) :version \\(.*\\)" s)
+                   `(,(match-string 1 s) . ,(match-string 2 s))))
+               res))))
+
+         (exe (if-let ((exe (executable-find "neil")))
+                  exe (user-error "Cannot find 'neil' command!")))
+
+         (res (funcall perform-action exe (concat "dep search " term)))
```
Comment (dakra) [inline neil.el:91]: Probably a good idea to use `shell-quote-argument` for `term`.

**Comment by agzam** [comment]
@borkdude Oh, I'm sorry. I honestly overlooked the headers. I just pushed the changes.

File: neil.el (hunk 2 above)
Comment (agzam) [reply neil.el]: Sorry, missed that.

File: neil.el (hunk 3 above)
Comment (agzam) [reply neil.el]: Thank you. I didn't know about shell-quote-argument. I learned something new today.

File: README.md (hunk 1 above)
Comment (agzam) [reply README.md]: Oh my. Allow me to blame my eyesight this time. Although, tbh, I was so excited to see it work, I cut the corners. Not very cool of me.

**Comment by agzam** [comment]
> I think adding to deps.edn should be enabled by default and can be optionally disabled. What else is the point for searching for a dep? :)

I thought about it, and it seemed that the default shouldn't be "destructive". I'm hoping to figure out different use-cases, like updating/changing the version of a dep-at-point, or grabbing all dependencies in the project and generating report of new versions. Later, I'll probably add additional commands or extend the main one.

I think we should keep the "main command" as simple as possible - once users start binding it to a key, they'd expect it not to break and behave the same way, unless the change the defaults.

One thing I am not sure, perhaps the name should be something like "neil-search-clj" or even more descriptive, so it's easily found in M-x menu?

**Comment by dakra** [comment]
> I thought about it, and it seemed that the default shouldn't be "destructive".

Imho `neil-search` sounds non-destructive. But I would simply make another command `neil-update-deps` or so.

Personally I would also like to toggle the `prompt-for-version` with a prefix argument.
So in my imagined setup I would call `neil-update-deps` to update my deps.clj file with the latest version
and when called with C-u it asks for which version. (Maybe a defcustom can change that behaviour around)

**Comment by teodorlu** [comment]
> One thing I am not sure, perhaps the name should be something like "neil-search-clj" or even more descriptive, so it's easily found in M-x menu?

I like to err on "to verbose" for interactive M-x commands.

I've discovered a lot of Doom's functionality this way. I want to move a window left. I just know what I want. "move", "left". Not sure which package those are in. So I do just that.

image

Oh, that's nice. It worked! If I want more, I can read the docs.

image

In this case, I'd like to add a dependency. "clojure add dependency". We're using neil specifically, so perhaps `clojure-neil-add-dependency` makes sense. I tend to like `clojure` as the first word, so that I can browse Clojure stuff with `M-x ^clojure`.

---

Note - this is a bunch of anectodal stuff / opinions. I'm just really happy I'm soon going to be able to use this :)

**Comment by teodorlu** [comment]
> I thought about it, and it seemed that the default shouldn't be "destructive". I'm hoping to figure out different use-cases, like updating/changing the version of a dep-at-point, or grabbing all dependencies in the project and generating report of new versions. Later, I'll probably add additional commands or extend the main one.
>
> I think we should keep the "main command" as simple as possible - once users start binding it to a key, they'd expect it not to break and behave the same way, unless the change the defaults.

> Imho neil-search sounds non-destructive. But I would simply make another command neil-update-deps or so.

I'd also be most interested in adding a dependency. Because ... well, I want to add a dependency. Perhaps add some "dwim" (do what I mean) suffix?

image

Having one command for "add dependency" and another for "search for dependencies" is possible. But I'm not sure when I'd just be searching for dependencies without doing anything with it. Perhaps "add dependency" is the interactive command, and "search for dependencies" is just a function that "add dependency" uses?

**Comment by teodorlu** [comment]
It could also be really interesting with an interactive buffer for a deps.edn file. Like `ibuffer`, just listing all the dependencies. Possibly also knowing about the difference between `:deps` and dependencies in `:aliases`.

But I don't know _how_ I'd use that yet. Perhaps adding functionality one piece at the time when it "makes sense" is good.

**Comment by teodorlu** [comment]
Apologies for the walls of text - a lot of this should probably be addressed in future pull requests. Thanks again for making this :)

**Comment by borkdude** [comment]
I'll wait for the signal for when I can merge this as there is still a discussion going on :)

**Comment by teodorlu** [comment]
I'm not opposed to mergning if @agzam wants to merge!

**Comment by agzam** [comment]
> Personally I would also like to toggle the `prompt-for-version` with a prefix argument. 

Exactly my thoughts, I wanted to add this feature but decided to keep the initial version streamlined. Once we figure out different use cases, I'll extend the functionality.

**Comment by agzam** [comment]
> so perhaps M-x clojure-neil-add-dependency makes sense.

@teodorlu I prefer to keep the names prefixed with the group-name, in your case it would be something like `neil-add-clojure-dependency`. I think I'm gonna rename the "entry-point" function to `neil-find-clojure-package`. And then we can merge this.

**Pull request merged**